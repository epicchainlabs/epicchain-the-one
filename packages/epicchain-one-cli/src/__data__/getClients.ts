import { getClients as getClientsBase } from '@neo-one/cli-common';
import { common, crypto } from '@neo-one/client-common';
import { NEOONEDataProvider } from '@neo-one/client-core';
import { constants } from '@neo-one/utils';

export const getClients = async (project: string) => {
  crypto.addPublicKey(
    common.stringToPrivateKey(constants.PRIVATE_NET_PRIVATE_KEY),
    common.stringToECPoint(constants.PRIVATE_NET_PUBLIC_KEY),
  );

  const config = await one.getProjectConfig(project);

  return getClientsBase(
    new NEOONEDataProvider({
      network: constants.LOCAL_NETWORK_NAME,
      rpcURL: `http://localhost:${config.network.port}/rpc`,
    }),
    constants.PRIVATE_NET_PRIVATE_KEY,
  );
};
