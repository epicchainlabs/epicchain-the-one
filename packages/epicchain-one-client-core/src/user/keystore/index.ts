export { LocalKeyStore, LocalWallet, LockedWallet, UnlockedWallet, LocalStore } from './LocalKeyStore';
export { HDKeyStore, HDAccount, HDProvider, HDHandler } from './HDKeyStore';
export * from './LocalMemoryStore';
export * from './LocalStringStore';
export * from './LocalHDStore';
export * from './LocalHDHandler';
export * from './createLedgerKeyStore';
export * from './LedgerHandler';
export * from './LedgerNanoStore';
export * from './types';
