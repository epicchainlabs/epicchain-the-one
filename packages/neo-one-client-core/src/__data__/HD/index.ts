export * from './seedOne';
export * from './seedThree';
export * from './seedTwo';
export * from './utils';
export * from './ledgerData';
