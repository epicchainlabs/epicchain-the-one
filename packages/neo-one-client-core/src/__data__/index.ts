export * from './keys';
export * from './data';
export * from './factory';
export * from './HD';
export * from './verifyDataProvider';
