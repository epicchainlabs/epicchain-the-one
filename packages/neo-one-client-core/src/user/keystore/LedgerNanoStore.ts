// wallaby.skip
import { crypto, PublicKeyString } from '@neo-one/client-common';
import { HWLedger, LedgerTransport } from '@neo-one/client-switch';
import {
  LedgerMessageSizeError,
  LedgerNEOAppError,
  LedgerNotDetectedError,
  LedgerNotSupportedError,
  LedgerParseError,
  LedgerStatusCodeError,
  LedgerTransactionDenied,
} from '../../errors';
import { HDStore } from './types';

export interface LedgerQueueRequest {
  readonly msg: Buffer;
  readonly resolve: (value: Buffer) => void;
  readonly reject: (error: Error) => void;
}

const keyRequestParams = Buffer.from('80040000', 'hex');
const initMessageParams = Buffer.from('80020000', 'hex');
const finalMessageParams = Buffer.from('80028000', 'hex');

const BIP44 = (accountIn = 0): Buffer => {
  const account = accountIn.toString(16);

  return Buffer.from(`8000002C800003788000000000000000${'0'.repeat(8 - account.length)}${account}`, 'hex');
};

const chunkBuffer = (buffer: Buffer, chunkLength: number): readonly Buffer[] => {
  if (buffer.length > chunkLength) {
    return [buffer.slice(0, chunkLength)].concat(chunkBuffer(buffer.slice(chunkLength), chunkLength));
  }

  return [buffer];
};

const processLedgerStatus = (status: string): void => {
  switch (status) {
    case '6d08': {
      throw new LedgerMessageSizeError();
    }
    case '6e00': {
      throw new LedgerNEOAppError();
    }
    case '6985': {
      throw new LedgerTransactionDenied();
    }
    case '6d07': {
      throw new LedgerParseError();
    }
    default: {
      throw new LedgerStatusCodeError(status);
    }
  }
};

export class LedgerNanoStore implements HDStore<number> {
  public static readonly byteLimit = LedgerTransport.byteLimit;
  public static readonly type = LedgerTransport.type;

  public static readonly init = async () => {
    const transport = await LedgerTransport.load();
    const isSupported = await transport.isSupported();
    if (!isSupported) {
      throw new LedgerNotSupportedError();
    }
    const paths = await transport.list();
    if (paths.length === 0) {
      throw new LedgerNotDetectedError();
    }
    const ledger = await transport.open(paths[0]);
    ledger.setScrambleKey('NEO');

    return new LedgerNanoStore(ledger);
  };

  private readonly ledger: HWLedger;
  private readonly mutableQueue: LedgerQueueRequest[];
  private mutableRunning: boolean;

  private constructor(ledger: HWLedger) {
    this.ledger = ledger;
    this.mutableRunning = false;
    this.mutableQueue = [];
  }

  public async getPublicKey(account = 0): Promise<PublicKeyString> {
    const response = await this.send(Buffer.concat([keyRequestParams, Buffer.from('14', 'hex'), BIP44(account)]));

    return crypto.toECPoint(response).toString('hex');
  }

  public async sign({ message, account }: { readonly message: Buffer; readonly account: number }): Promise<Buffer> {
    const data = Buffer.concat([message, BIP44(account)]);
    const chunks = chunkBuffer(data, 250);
    const initChunks = chunks.slice(0, chunks.length - 1);
    const finalChunk = chunks[chunks.length - 1];

    await Promise.all(
      initChunks.map(async (chunk) => {
        await this.send(Buffer.concat([initMessageParams, Buffer.from([chunk.length]), chunk]));
      }),
    );

    return this.send(
      Buffer.concat([finalMessageParams, Buffer.from(finalChunk.length.toString(16), 'hex'), finalChunk]),
    );
  }

  public async close(): Promise<void> {
    await this.ledger.close();
  }

  private async send(msg: Buffer): Promise<Buffer> {
    // tslint:disable-next-line:promise-must-complete
    const promise = new Promise<Buffer>((resolve, reject) => {
      this.mutableQueue.push({
        msg,
        resolve,
        reject,
      });
    });

    this.startSendLoop().catch(() => {
      // do nothing
    });

    return promise;
  }

  private async startSendLoop(): Promise<void> {
    if (this.mutableRunning) {
      return;
    }

    this.mutableRunning = true;

    let entry = this.mutableQueue.shift();
    // tslint:disable-next-line:no-loop-statement
    while (entry !== undefined) {
      try {
        const result = await this.doSend(entry.msg);
        entry.resolve(result);
      } catch (error) {
        entry.reject(error);
      }
      entry = this.mutableQueue.shift();
    }

    this.mutableRunning = false;
  }

  private async doSend(msg: Buffer): Promise<Buffer> {
    const response = await this.ledger.exchange(msg);

    const status = response.slice(response.length - 2).toString('hex');
    if (status !== '9000') {
      processLedgerStatus(status);
    }

    return response.slice(0, response.length - 2);
  }
}
