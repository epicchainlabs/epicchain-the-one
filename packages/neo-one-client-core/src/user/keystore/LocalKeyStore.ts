import {
  AccountContract,
  BufferString,
  common,
  crypto,
  decryptNEP2,
  encryptNEP2,
  NetworkType,
  PrivateKeyString,
  privateKeyToPublicKey,
  publicKeyToAddress,
  UpdateAccountNameOptions,
  UserAccount,
  UserAccountID,
} from '@neo-one/client-common';
import { utils } from '@neo-one/utils';
import debug from 'debug';
import _ from 'lodash';
import { BehaviorSubject, Observable } from 'rxjs';
import { distinctUntilChanged, map } from 'rxjs/operators';
import * as args from '../../args';
import { LockedAccountError, UnknownAccountError } from '../../errors';
import { KeyStore } from '../LocalUserAccountProvider';

const logger = debug('NEOONE:LocalKeystore');

/**
 * Wallet in the "locked" state.
 */
export interface LockedWallet {
  /**
   * `type` differentiates a `LockedWallet` from other `LocalWallet`s, i.e. an `UnlockedWallet`
   */
  readonly type: 'locked';
  /**
   * `UserAccount` this `LockedWallet` refers to.
   */
  readonly userAccount: UserAccount;
  /**
   * NEP-2 encrypted key of this `LockedWallet`.
   */
  readonly nep2: string;
}

/**
 * Wallet in the "unlocked" state.
 */
export interface UnlockedWallet {
  /**
   * `type` differentiates an `UnlockedWallet` from other `LocalWallet`s, i.e. an `LockedWallet`
   */
  readonly type: 'unlocked';
  /**
   * `UserAccount` this `UnlockedWallet` refers to.
   */
  readonly userAccount: UserAccount;
  /**
   * Private key for this `UnlockedWallet`.
   */
  readonly privateKey: PrivateKeyString;
  /**
   * NEP-2 encrypted key of this `UnlockedWallet`. `undefined` if the `privateKey` has never been encrypted.
   */
  readonly nep2?: string | undefined;
}

/**
 * Locally stored wallet that is either in a `'locked'` or `'unlocked'` state (`type`).
 */
export type LocalWallet = LockedWallet | UnlockedWallet;
/**
 * Mapping from `NetworkType` -> `AddressString` -> `LocalWallet`
 */
export type Wallets = { readonly [Network in string]?: { readonly [Address in string]?: LocalWallet } };

/**
 * Base interface that must be implemented to use the `LocalKeyStore`.
 */
export interface LocalStore {
  /**
   * @returns `Promise` that resolves to all available `LocalWallet`s.
   */
  readonly getWallets: () => Promise<readonly LocalWallet[]>;
  /**
   * Optional method that returns the available wallets synchronously.
   *
   * @returns All available `LocalWallet`s
   */
  readonly getWalletsSync?: () => readonly LocalWallet[];
  /**
   * Save a wallet to the store.
   */
  readonly saveWallet: (wallet: LocalWallet) => Promise<void>;
  /**
   * Delete a wallet from the store.
   */
  readonly deleteWallet: (account: LocalWallet) => Promise<void>;
}

const flattenWallets = (wallets: Wallets) =>
  _.flatten(
    Object.values(wallets)
      .filter(utils.notNull)
      .map((networkWallets) => Object.values(networkWallets)),
  ).filter(utils.notNull);

/**
 * `LocalKeyStore` implements the `KeyStore` interface expected by `LocalUserAccountProvider` via an underlying `Store` implementation.
 */
export class LocalKeyStore implements KeyStore {
  public readonly currentUserAccount$: Observable<UserAccount | undefined>;
  public readonly userAccounts$: Observable<readonly UserAccount[]>;
  public readonly wallets$: Observable<readonly LocalWallet[]>;
  private readonly currentAccountInternal$: BehaviorSubject<UserAccount | undefined>;
  private readonly accountsInternal$: BehaviorSubject<readonly UserAccount[]>;
  private readonly walletsInternal$: BehaviorSubject<Wallets>;
  private readonly store: LocalStore;
  private readonly initPromise: Promise<void>;

  public constructor(store: LocalStore) {
    this.walletsInternal$ = new BehaviorSubject<Wallets>({});
    this.wallets$ = this.walletsInternal$.pipe(
      distinctUntilChanged((a, b) => _.isEqual(a, b)),
      map(flattenWallets),
    );

    this.accountsInternal$ = new BehaviorSubject([] as readonly UserAccount[]);
    this.wallets$
      .pipe(map((wallets) => wallets.map(({ userAccount }) => userAccount)))
      .subscribe(this.accountsInternal$);
    this.userAccounts$ = this.accountsInternal$;

    this.currentAccountInternal$ = new BehaviorSubject(undefined as UserAccount | undefined);
    this.currentUserAccount$ = this.currentAccountInternal$.pipe(distinctUntilChanged());

    this.store = store;

    if (store.getWalletsSync !== undefined) {
      this.initWithWallets(store.getWalletsSync());
    }
    this.initPromise = this.init();
  }

  public getCurrentUserAccount(): UserAccount | undefined {
    return this.currentAccountInternal$.getValue();
  }

  public getUserAccounts(): readonly UserAccount[] {
    return this.accountsInternal$.getValue();
  }

  public get wallets(): readonly LocalWallet[] {
    return flattenWallets(this.walletsObj);
  }

  public async sign({
    account,
    message,
  }: {
    readonly account: UserAccountID;
    readonly message: string;
  }): Promise<string> {
    await this.initPromise;

    return this.capture(async () => {
      const privateKey = this.getPrivateKey(account);

      return crypto
        .sign({ message: Buffer.from(message, 'hex'), privateKey: common.stringToPrivateKey(privateKey) })
        .toString('hex');
    }, 'neo_sign');
  }

  public async selectUserAccount(id?: UserAccountID): Promise<void> {
    if (id === undefined) {
      this.currentAccountInternal$.next(undefined);
      this.newCurrentAccount();
    } else {
      const { userAccount } = this.getWallet(id);
      this.currentAccountInternal$.next(userAccount);
    }
  }

  public async updateUserAccountName({ id, name }: UpdateAccountNameOptions): Promise<void> {
    return this.capture(async () => {
      await this.initPromise;

      const wallet = this.getWallet(id);
      let newWallet: LocalWallet;
      const userAccount = {
        id: wallet.userAccount.id,
        name,
        publicKey: wallet.userAccount.publicKey,
        contract: wallet.userAccount.contract,
      };

      if (wallet.type === 'locked') {
        newWallet = {
          type: 'locked',
          userAccount,
          nep2: wallet.nep2,
        };
      } else {
        newWallet = {
          type: 'unlocked',
          userAccount,
          privateKey: wallet.privateKey,
          nep2: wallet.nep2,
        };
      }
      await this.capture(async () => this.store.saveWallet(newWallet), 'neo_save_wallet');

      this.updateWallet(newWallet);
    }, 'neo_update_account_name');
  }

  public getWallet({ address, network }: UserAccountID): LocalWallet {
    const wallets = this.walletsObj[network];
    if (wallets === undefined) {
      throw new UnknownAccountError(address);
    }

    const wallet = wallets[address];
    if (wallet === undefined) {
      throw new UnknownAccountError(address);
    }

    return wallet;
  }

  public getWallet$({ address, network }: UserAccountID): Observable<LocalWallet | undefined> {
    return this.walletsInternal$.pipe(
      map((wallets) => {
        const networkWallets = wallets[network];
        if (networkWallets === undefined) {
          return undefined;
        }

        return networkWallets[address];
      }),
    );
  }

  public async addUserAccount({
    network,
    privateKey: privateKeyIn,
    name,
    password,
    nep2: nep2In,
  }: {
    readonly network: NetworkType;
    readonly privateKey?: BufferString;
    readonly name?: string;
    readonly password?: string;
    readonly nep2?: string;
  }): Promise<LocalWallet> {
    await this.initPromise;

    return this.capture(async () => {
      let pk = privateKeyIn;
      let nep2 = nep2In;
      if (pk === undefined) {
        if (nep2 === undefined || password === undefined) {
          throw new Error('Expected private key or password and NEP-2 key');
        }
        pk = await decryptNEP2({ encryptedKey: nep2, password });
      }

      const privateKey = args.assertPrivateKey('privateKey', pk);
      const publicKey = privateKeyToPublicKey(privateKey);
      const address = publicKeyToAddress(publicKey);

      if (nep2 === undefined && password !== undefined) {
        nep2 = await encryptNEP2({ privateKey, password });
      }

      const contract = AccountContract.createSignatureContract(common.stringToECPoint(publicKey));

      const userAccount = {
        id: {
          network,
          address,
        },
        name: name === undefined ? address : name,
        publicKey,
        contract,
      };

      const unlockedWallet: UnlockedWallet = {
        type: 'unlocked',
        userAccount,
        nep2,
        privateKey,
      };

      let wallet: LocalWallet = unlockedWallet;
      if (nep2 !== undefined) {
        wallet = { type: 'locked', userAccount, nep2 };
      }

      await this.capture(async () => this.store.saveWallet(wallet), 'neo_save_wallet');

      this.updateWallet(unlockedWallet);

      if (this.currentAccount === undefined) {
        this.currentAccountInternal$.next(wallet.userAccount);
      }

      return unlockedWallet;
    }, 'neo_add_account');
  }

  /**
   * not a true implementation, just serves as a helper for private net.
   * @internal
   */
  public async addMultiSigUserAccount({
    network,
    privateKeys: privateKeysIn,
    name,
  }: {
    readonly network: NetworkType;
    readonly privateKeys: readonly BufferString[];
    readonly name?: string;
  }): Promise<LocalWallet> {
    await this.initPromise;

    return this.capture(async () => {
      const pk = privateKeysIn[0];

      const privateKey = args.assertPrivateKey('privateKey', pk);
      const publicKey = privateKeyToPublicKey(privateKey);
      const ecPoint = common.stringToECPoint(publicKey);
      const address = crypto.scriptHashToAddress({
        addressVersion: common.NEO_ADDRESS_VERSION,
        scriptHash: crypto.toScriptHash(crypto.createMultiSignatureVerificationScript(1, [ecPoint])),
      });

      const contract = AccountContract.createMultiSigContract(1, [ecPoint]);

      const userAccount = {
        id: {
          network,
          address,
        },
        name: name === undefined ? address : name,
        publicKey,
        contract,
      };

      const unlockedWallet: UnlockedWallet = {
        type: 'unlocked',
        userAccount,
        privateKey,
      };

      const wallet: LocalWallet = unlockedWallet;

      await this.capture(async () => this.store.saveWallet(wallet), 'neo_save_wallet');

      this.updateWallet(unlockedWallet);

      if (this.currentAccount === undefined) {
        this.currentAccountInternal$.next(wallet.userAccount);
      }

      return unlockedWallet;
    }, 'neo_add_account');
  }

  public async deleteUserAccount(id: UserAccountID): Promise<void> {
    await this.initPromise;

    return this.capture(async () => {
      const { walletsObj: wallets } = this;
      const networkWalletsIn = wallets[id.network];
      if (networkWalletsIn === undefined) {
        return;
      }

      const { [id.address]: wallet, ...networkWallets } = networkWalletsIn;
      if (wallet === undefined) {
        return;
      }

      await this.capture(async () => this.store.deleteWallet(wallet), 'neo_delete_wallet');

      this.walletsInternal$.next({
        ...wallets,
        [id.network]: networkWallets,
      });

      if (
        this.currentAccount !== undefined &&
        this.currentAccount.id.network === id.network &&
        this.currentAccount.id.address === id.address
      ) {
        this.newCurrentAccount();
      }
    }, 'neo_delete_account');
  }

  public async unlockWallet({
    id,
    password,
  }: {
    readonly id: UserAccountID;
    readonly password: string;
  }): Promise<void> {
    await this.initPromise;

    return this.capture(async () => {
      const wallet = this.getWallet(id);
      if (wallet.type === 'unlocked') {
        return;
      }

      const privateKey = await decryptNEP2({
        encryptedKey: wallet.nep2,
        password,
      });

      this.updateWallet({
        type: 'unlocked',
        userAccount: wallet.userAccount,
        privateKey,
        nep2: wallet.nep2,
      });
    }, 'neo_unlock_wallet');
  }

  public async lockWallet(id: UserAccountID): Promise<void> {
    await this.initPromise;

    const wallet = this.getWallet(id);
    if (wallet.type === 'locked' || wallet.nep2 === undefined) {
      return;
    }

    this.updateWallet({
      type: 'locked',
      userAccount: wallet.userAccount,
      nep2: wallet.nep2,
    });
  }

  private async init(): Promise<void> {
    const walletsList = await this.store.getWallets();
    this.initWithWallets(walletsList);
  }

  private initWithWallets(walletsList: readonly LocalWallet[]): void {
    const wallets = walletsList.reduce<Wallets>(
      (acc, wallet) => ({
        ...acc,
        [wallet.userAccount.id.network]: {
          ...(acc[wallet.userAccount.id.network] === undefined ? {} : acc[wallet.userAccount.id.network]),
          [wallet.userAccount.id.address]: wallet,
        },
      }),
      {},
    );
    this.walletsInternal$.next(wallets);
    this.newCurrentAccount();
  }

  private getPrivateKey(id: UserAccountID): BufferString {
    const wallet = this.getWallet({
      network: id.network,
      address: id.address,
    });

    if (wallet.type === 'locked') {
      throw new LockedAccountError(id.address);
    }

    return wallet.privateKey;
  }

  private updateWallet(wallet: LocalWallet): void {
    const { walletsObj: wallets } = this;
    this.walletsInternal$.next({
      ...wallets,
      [wallet.userAccount.id.network]: {
        ...(wallets[wallet.userAccount.id.network] === undefined ? {} : wallets[wallet.userAccount.id.network]),
        [wallet.userAccount.id.address]: wallet,
      },
    });
  }

  private newCurrentAccount(): void {
    const allAccounts = this.wallets.map(({ userAccount: value }) => value);

    const account = allAccounts[0];
    this.currentAccountInternal$.next(account);
  }

  private async capture<T>(func: () => Promise<T>, title: string): Promise<T> {
    try {
      const result = await func();
      logger('%o', { title, level: 'debug' });

      return result;
    } catch (error) {
      logger('%o', { title, level: 'error', error: error.message });
      throw error;
    }
  }

  private get walletsObj(): Wallets {
    return this.walletsInternal$.getValue();
  }

  private get currentAccount(): UserAccount | undefined {
    return this.currentAccountInternal$.getValue();
  }
}
