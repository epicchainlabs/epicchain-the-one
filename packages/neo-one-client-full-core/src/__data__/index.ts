export * from './keys';
export * from './data';
export * from './factory';
