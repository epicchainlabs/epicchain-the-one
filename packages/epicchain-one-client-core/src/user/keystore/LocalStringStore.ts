import { AccountContract, UserAccountJSON } from '@neo-one/client-common';
import { PasswordRequiredError } from '../../errors';
import { LocalStore, LocalWallet, LockedWallet, UnlockedWallet } from './LocalKeyStore';

/**
 * Interface that `LocalStringStore` requires to function.
 */
export interface LocalStringStoreStorage {
  /**
   * Set `key` to `value`.
   */
  readonly setItem: (key: string, value: string) => Promise<void>;
  /**
   * Return the value of `key`
   */
  readonly getItem: (key: string) => Promise<string>;
  /**
   * Remove `key`.
   */
  readonly removeItem: (key: string) => Promise<void>;
  /**
   * Return all keys.
   */
  readonly getAllKeys: () => Promise<readonly string[]>;
}

export interface LockedWalletJSON extends Omit<LockedWallet, 'userAccount'> {
  readonly userAccount: UserAccountJSON;
}

export interface UnlockedWalletJSON extends Omit<UnlockedWallet, 'userAccount'> {
  readonly userAccount: UserAccountJSON;
}

export type LocalWalletJSON = LockedWalletJSON | UnlockedWalletJSON;

const toWalletJSON = (wallet: LocalWallet): LocalWalletJSON => {
  const { contract, ...rest } = wallet.userAccount;
  const userAccountJSON = {
    ...rest,
    contract: contract.serializeJSON(),
  };
  if (wallet.userAccount.id.network === 'main') {
    if (wallet.nep2 === undefined) {
      throw new PasswordRequiredError();
    }

    return {
      type: 'locked',
      userAccount: userAccountJSON,
      nep2: wallet.nep2,
    };
  }

  const { userAccount: _userAccount, ...walletRest } = wallet;

  return {
    ...walletRest,
    userAccount: userAccountJSON,
  };
};

const fromWalletJSON = (wallet: LocalWalletJSON): LocalWallet => {
  const {
    userAccount: { contract, ...accountRest },
    ...rest
  } = wallet;
  const deserializedContract = AccountContract.fromJSON(contract);

  const userAccount = {
    ...accountRest,
    contract: deserializedContract,
  };

  return {
    ...rest,
    userAccount,
  };
};

/**
 * Implements the `LocalStore` interface expected by `LocalKeyStore`.
 */
export class LocalStringStore implements LocalStore {
  public constructor(public readonly storage: LocalStringStoreStorage) {}

  public async getWallets(): Promise<readonly LocalWallet[]> {
    const keys = await this.storage.getAllKeys();
    const values = await Promise.all(keys.map(async (key) => this.storage.getItem(key)));

    return values.map((value) => fromWalletJSON(JSON.parse(value)));
  }

  public async saveWallet(wallet: LocalWallet): Promise<void> {
    let safeWallet = wallet;
    if (wallet.userAccount.id.network === 'main') {
      if (wallet.nep2 === undefined) {
        throw new PasswordRequiredError();
      }
      safeWallet = {
        type: 'locked',
        userAccount: wallet.userAccount,
        nep2: wallet.nep2,
      };
    }

    await this.storage.setItem(this.getKey(safeWallet), JSON.stringify(toWalletJSON(safeWallet)));
  }

  public async deleteWallet(wallet: LocalWallet): Promise<void> {
    await this.storage.removeItem(this.getKey(wallet));
  }

  private getKey({
    userAccount: {
      id: { network, address },
    },
  }: LocalWallet): string {
    return `${network}-${address}`;
  }
}
