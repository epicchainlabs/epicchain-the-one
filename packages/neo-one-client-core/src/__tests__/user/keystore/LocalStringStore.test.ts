import { factory } from '../../../__data__';
import { LocalStringStore } from '../../../user/keystore/LocalStringStore';

describe('LocalStringStore', () => {
  const createStorage = () => {
    const mutableItems: { [key: string]: string } = {};

    return {
      setItem: async (key: string, value: string) => {
        mutableItems[key] = value;
      },
      getItem: async (key: string) => {
        const value = mutableItems[key] as string | undefined;
        if (value === undefined) {
          throw new Error(`Key not found: ${key}`);
        }

        return value;
      },
      removeItem: async (key: string) => {
        // tslint:disable-next-line no-dynamic-delete
        delete mutableItems[key];
      },
      getAllKeys: async () => Object.keys(mutableItems),
    };
  };

  let store: LocalStringStore;
  beforeEach(() => {
    store = new LocalStringStore(createStorage());
  });

  test('getWallets', async () => {
    const result = await store.getWallets();

    expect(result).toHaveLength(0);
  });

  test('saveWallet', async () => {
    const wallet = factory.createUnlockedWallet();

    await store.saveWallet(wallet);

    const result = await store.getWallets();
    expect(result).toHaveLength(1);
    expect(result[0].type).toEqual('locked');
    expect(result[0].userAccount).toEqual(wallet.userAccount);
    expect(result[0].nep2).toEqual(wallet.nep2);
  });

  test('saveWallet - locked wallet', async () => {
    const wallet = factory.createLockedWallet();

    await store.saveWallet(wallet);

    const result = await store.getWallets();
    expect(result).toHaveLength(1);
    expect(result[0].type).toEqual('locked');
    expect(result[0].userAccount).toEqual(wallet.userAccount);
    expect(result[0].nep2).toEqual(wallet.nep2);
  });

  test('saveWallet - nep2 undefined', async () => {
    const { nep2: _unused, ...wallet } = factory.createUnlockedWallet();

    const result = store.saveWallet(wallet);

    await expect(result).rejects.toMatchSnapshot();
  });

  test('deleteWallet', async () => {
    const wallet = factory.createUnlockedWallet();

    await store.saveWallet(wallet);
    await store.deleteWallet(wallet);

    const result = await store.getWallets();
    expect(result).toHaveLength(0);
  });
});
