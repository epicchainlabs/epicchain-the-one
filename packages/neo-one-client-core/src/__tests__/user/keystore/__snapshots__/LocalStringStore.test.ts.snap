// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocalStringStore saveWallet - nep2 undefined 1`] = `[Error [PASSWORD_REQUIRED]: A password is required when creating accounts on the MainNet.]`;
